Promote Style to public API

jfirebaugh opened the pull request:
Make `style.hpp` a public header, provide `Style& Map::getStyle()`, and move runtime-styling APIs that previously lived on `Map` to `Style`. This is a refactor made relatively easy by the async rendering changes. It reduces the responsibilities of the `Map` object itself and is a prerequisite for #6386. @1ec5, can you review what else is needed from core to support #6386? In particular, do we need `Map::setStyle(std::unique_ptr<Style>)`? Does `Style` need to be constructible without the `Scheduler&, FileSource&, float pixelRatio` arguments? I think we can hide those if need be.

1ec5 commented on include/mbgl/map/map.hpp:
```diff
@@ -58,15 +55,15 @@ class Map : private util::noncopyable {
-    void setTransitionOptions(const style::TransitionOptions&);
-
+    // Style
     void setStyleURL(const std::string&);
```
With this method, the workflow is to create the map, then set its style to a URL. #6386 also proposes exposing an  `-[MGLStyle initWithURL:]`. I think this would require a way to initialize an `mbgl::style::Style` with nothing but a URL, so that later on `mbgl::Map::setStyle()` could be called, passing in that style. Do you see any complications with that approach? `-[MGLStyle initWithJSONData:]` (analogous to `setStyleJSON()`) is also a highly requested feature.

1ec5 commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -2152,10 +2153,10 @@ - (void)resetNorthAnimated:(BOOL)animated
+    CGFloat pitch = _mbglMap->getStyle().getDefaultPitch();
+    CLLocationDirection heading = mbgl::util::wrap(_mbglMap->getStyle().getDefaultBearing(), 0., 360.);
+    CLLocationDistance distance = MGLAltitudeForZoomLevel(_mbglMap->getStyle().getDefaultZoom(), pitch, 0, self.frame.size);
+    self.camera = [MGLMapCamera cameraLookingAtCenterCoordinate:MGLLocationCoordinate2DFromLatLng(_mbglMap->getStyle().getDefaultLatLng())
```
Ultimately, we’d add more internal properties to MGLStyle so that calls to `_mbglMap->getStyle()` become largely unnecessary in MGLMapView’s implementation.

jfirebaugh commented:
Cool -- I moved the JSON and URL loading methods from `Map` to `Style` as well, and added `Map::setStyle`.

> I’m less sure about `Scheduler` – it would be great if we could hide that parameter.

Now that the loading methods are on `Style`, these parameters are all necessary. SDKs can pass the same `Scheduler&` as they pass to the `Map` constructor.

ivovandongen commented on src/mbgl/map/map.cpp:
```diff
@@ -166,14 +153,12 @@ Map::Impl::Impl(Map& map_,
           }
       }) {
     style = std::make_unique<Style>(scheduler, fileSource, pixelRatio);
```
Not part of this pr. But why aren't we initialising this in the member initialisation list?

jfirebaugh replied:
It probably should be.

ivovandongen commented on include/mbgl/style/style.hpp:
```diff
@@ -0,0 +1,81 @@
+    void setTransitionOptions(const TransitionOptions&);
+
+    // Light
+          Light* getLight();
```
Should we use indentation here that is more common? Not sure if this helps with readability, but it prevents using standard formatting tools.

jfirebaugh replied:
It make clear the parallelism with the `const` version of the accessor -- you can immediately see where the `const`s are added, that they are in the right place, and that there aren't any other differences in the signatures. I'm opposed to using formatting tools that can't handle nuances like this.

ivovandongen commented on src/mbgl/style/style_impl.cpp:
```diff
@@ -0,0 +1,372 @@
+    loaded = false;
+    url = url_;
+
+    styleRequest = nullptr;
```
Is this necessary? The next line would already destruct any previous style before assignment right?

jfirebaugh replied:
It's not necessary, I'll remove it.

ivovandongen commented on src/mbgl/style/style_impl.cpp:
```diff
@@ -0,0 +1,372 @@
+}
+
+Layer* Style::Impl::addLayer(std::unique_ptr<Layer> layer, optional<std::string> before) {
+    // TODO: verify source
```
Great you got rid of BackendScope here!
